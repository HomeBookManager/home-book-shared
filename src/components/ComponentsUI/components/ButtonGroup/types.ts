// types
import { TButtonProps } from '../Button/Button';

export type TButtonGroupPickedButtonProps = Pick<
  TButtonProps,
  | 'className'
  | 'color'
  | 'disabled'
  | 'disabledRippleEffect'
  | 'e2eAttribute'
  | 'e2eValue'
  | 'forcedHover'
  | 'size'
  | 'style'
  | 'type'
  | 'variant'
>;

export type TButtonGroupContext = TButtonGroupPickedButtonProps | undefined;
