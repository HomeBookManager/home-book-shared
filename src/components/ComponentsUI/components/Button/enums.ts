export enum ButtonColor {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  warning = 'warning',
  error = 'error',
}

export enum ButtonSize {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

export enum ButtonVariant {
  text = 'text',
  contained = 'contained',
  outlined = 'outlined',
}
