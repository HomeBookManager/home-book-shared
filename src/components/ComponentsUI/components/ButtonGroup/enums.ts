export enum ButtonGroupOrientation {
  horizontal = 'horizontal',
  vertical = 'vertical',
}
