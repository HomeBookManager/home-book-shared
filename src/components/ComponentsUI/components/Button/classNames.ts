import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';

export const className = 'Button';

export const getIconModificators = (property: 'fill' | 'stroke') => ({
  ...mapValues(ButtonColor, (color) => ({
    ...mapValues(
      ButtonVariant,
      (variant) =>
        `${className}__icon--apply-${property}--${variant}--${color}`,
    ),
  })),
});

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      fullwidth: `${className}--full-width`,
      forcedHover: {
        ...mapValues(ButtonVariant, (variant) => ({
          ...mapValues(
            ButtonColor,
            (color) => `${className}--${variant}--${color}--forced-hover`,
          ),
        })),
      },
      ...mapValues(ButtonColor, (color) => `${className}--${color}`),
      ...mapValues(ButtonSize, (size) => `${className}--${size}`),
      ...mapValues(ButtonVariant, (variant) => `${className}--${variant}`),
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      applyFill: `${className}__icon--apply-fill`,
      applyStroke: `${className}__icon--apply-stroke`,
      end: `${className}__icon--end`,
      start: `${className}__icon--start`,
      ...mapValues(ButtonSize, (size) => `${className}__icon--${size}`),
    },
  },
};
