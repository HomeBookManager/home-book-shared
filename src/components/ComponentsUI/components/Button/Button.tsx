import cx from 'classnames';
import { ButtonHTMLAttributes, ReactNode, Ref, forwardRef } from 'react';

// components
import E2EDataAttribute from '../../../E2EDataAttributes/E2EDataAttribute';

// hooks
import { useButtonGroupContext } from './hooks/useButtonGroupContext';
import { useClickInteraction } from './hooks/useClickInteraction';
import { useIcon } from './hooks/useIcon';
import { useRippleEffect } from '../../../../hooks/useRippleEffect/useRippleEffect';

// others
import { className as classNameButton, classNames } from './classNames';

// styles
import './styles/button.scss';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import { TButtonIcon } from './types';

export type TButtonProps = Omit<
  ButtonHTMLAttributes<HTMLElement>,
  'disabled'
> & {
  children?: ReactNode;
  className?: string;
  color?: ButtonColor;
  disabled?: boolean;
  disabledRippleEffect?: boolean;
  e2eAttribute?: string;
  e2eValue?: number | string;
  endIcon?: TButtonIcon;
  forcedHover?: boolean;
  fullWidth?: boolean;
  onClick?: () => void;
  ref?: Ref<HTMLButtonElement>;
  size?: ButtonSize;
  startIcon?: TButtonIcon;
  variant?: ButtonVariant;
};

export const Button = forwardRef<HTMLButtonElement, TButtonProps>(
  (
    {
      children,
      className = '',
      color = ButtonColor.primary,
      disabled = false,
      disabledRippleEffect = false,
      endIcon = null,
      e2eAttribute = E2EAttribute.button,
      e2eValue = '',
      forcedHover = false,
      fullWidth = false,
      onClick,
      size = ButtonSize.medium,
      startIcon = null,
      type = 'button',
      variant = ButtonVariant.contained,
      ...restProps
    },
    ref,
  ) => {
    const {
      classNameButtonGroup,
      colorButtonGroup,
      disabledRippleEffectButtonGroup,
      forcedHoverButtonGroup,
      sizeButtonGroup,
      variantButtonGroup,
      ...restButtonGroupProps
    } = useButtonGroupContext();

    const Icon = useIcon(size);

    const { rippleEffect, triggerRippleEffect } = useRippleEffect(
      classNames[classNameButton].name,
    );

    const onClickHandler = useClickInteraction(
      disabledRippleEffect || disabledRippleEffectButtonGroup,
      onClick,
      triggerRippleEffect,
    );

    return (
      <E2EDataAttribute type={e2eAttribute} value={e2eValue}>
        <button
          className={cx(
            className,
            classNameButtonGroup,
            classNames[classNameButton].name,
            { [classNames[classNameButton].modificators.fullwidth]: fullWidth },
            {
              [classNames[classNameButton].modificators.forcedHover]:
                forcedHover || forcedHoverButtonGroup,
            },
            classNames[classNameButton].modificators[colorButtonGroup || color],
            classNames[classNameButton].modificators[sizeButtonGroup || size],
            classNames[classNameButton].modificators[
              variantButtonGroup || variant
            ],
          )}
          disabled={disabled}
          onClick={onClickHandler}
          ref={ref}
          type={type}
          {...restButtonGroupProps}
          {...restProps}
        >
          {startIcon && <Icon placement="start" src={startIcon} />}
          {children}
          {endIcon && <Icon placement="end" src={endIcon} />}
          {rippleEffect}
        </button>
      </E2EDataAttribute>
    );
  },
);

export default Button;
