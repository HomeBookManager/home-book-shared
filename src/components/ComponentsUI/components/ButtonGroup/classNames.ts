import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';

export const className = 'ButtonGroup';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      ...mapValues(
        ButtonGroupOrientation,
        (orientation) => `${className}--${orientation}`,
      ),
    },
  },
  button: {
    name: `${className}__button`,
    modificators: {
      forcedHover: `${className}__button--forced-hover`,
      ...mapValues(ButtonColor, (color) => `${className}__button--${color}`),
      ...mapValues(
        ButtonGroupOrientation,
        (orientation) => `${className}__button--${orientation}`,
      ),
      ...mapValues(
        ButtonVariant,
        (variant) => `${className}__button--${variant}`,
      ),
    },
  },
};
