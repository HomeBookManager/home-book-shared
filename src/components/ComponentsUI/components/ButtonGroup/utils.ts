import cx from 'classnames';

// others
import { classNames } from './classNames';

// types
import { ButtonColor, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';

export const composeButtonClassNames = (
  classNameButton: string,
  color: ButtonColor,
  forcedHover: boolean,
  orientation: ButtonGroupOrientation,
  variant: ButtonVariant,
): string =>
  cx(
    classNameButton,
    classNames.button.name,
    classNames.button.modificators[color],
    {
      [classNames.button.modificators.forcedHover]: forcedHover,
    },
    classNames.button.modificators[orientation],
    classNames.button.modificators[variant],
  );
