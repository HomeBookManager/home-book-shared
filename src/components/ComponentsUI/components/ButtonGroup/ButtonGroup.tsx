import cx from 'classnames';
import {
  HTMLAttributes,
  ReactElement,
  Ref,
  createContext,
  forwardRef,
} from 'react';

// components
import E2EDataAttribute from '../../../E2EDataAttributes/E2EDataAttribute';

// others
import { className as classNameButtonGroup, classNames } from './classNames';

// styles
import './styles/button-group.scss';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import { TButtonGroupContext, TButtonGroupPickedButtonProps } from './types';
import { TButtonProps } from '../Button/Button';

// utils
import { composeButtonClassNames } from './utils';

export const ButtonGroupContext = createContext<TButtonGroupContext>(undefined);

export type TButtonGroupProps = HTMLAttributes<HTMLDivElement> &
  TButtonGroupPickedButtonProps & {
    children: ReactElement<TButtonProps> | Array<ReactElement<TButtonProps>>;
    className?: string;
    classNameButton?: string;
    orientation?: ButtonGroupOrientation;
    ref?: Ref<HTMLDivElement>;
  };

export const ButtonGroup = forwardRef<HTMLDivElement, TButtonGroupProps>(
  (
    {
      children,
      className = '',
      classNameButton = '',
      color = ButtonColor.primary,
      disabled = false,
      disabledRippleEffect = false,
      e2eAttribute = E2EAttribute.buttonGroup,
      e2eValue = '',
      forcedHover = false,
      orientation = ButtonGroupOrientation.horizontal,
      size = ButtonSize.medium,
      style = {},
      type = 'button',
      variant = ButtonVariant.contained,
      ...restProps
    },
    ref,
  ) => {
    const buttonClassNames = composeButtonClassNames(
      classNameButton,
      color,
      forcedHover,
      orientation,
      variant,
    );

    return (
      <E2EDataAttribute type={e2eAttribute} value={e2eValue}>
        <div
          className={cx(
            className,
            classNames[classNameButtonGroup].name,
            classNames[classNameButtonGroup].modificators[orientation],
          )}
          ref={ref}
          style={style}
          {...restProps}
        >
          <ButtonGroupContext.Provider
            value={{
              className: buttonClassNames,
              color,
              disabled,
              disabledRippleEffect,
              forcedHover,
              size,
              type,
              variant,
            }}
          >
            {children}
          </ButtonGroupContext.Provider>
        </div>
      </E2EDataAttribute>
    );
  },
);

export default ButtonGroup;
