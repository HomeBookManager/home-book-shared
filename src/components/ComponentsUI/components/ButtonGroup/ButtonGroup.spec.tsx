import { fireEvent, render, waitFor } from '@testing-library/react';

// components
import Button from '../Button/Button';
import ButtonGroup from './ButtonGroup';

// others
import { classNames as classNamesButtonGroup } from './classNames';
import {
  className as classNameButton,
  classNames as classNamesButton,
} from '../Button/classNames';
import { RIPPLE_EFFECT_MODIFICATOR } from '../../../../hooks';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';

// utils
import { getByE2EAttribute } from '../../../../tests/testHelpers';
import { getDataTestAttribute } from '../../../E2EDataAttributes/utils';
import { enumToArray } from '../../../../utils';

const contents = ['Button 1', 'Button 2', 'Button 3'];

const children = (
  <>
    {contents.map((content, index) => (
      <Button e2eValue={index} key={content}>
        {content}
      </Button>
    ))}
  </>
);

const className = 'className';

describe('ButtonGroup behaviors', () => {
  it('should render rippleEffect after click for every button', async () => {
    // before
    const { container } = render(<ButtonGroup>{children}</ButtonGroup>);

    for (let i = 0; i < contents.length; i++) {
      // find
      const button = getByE2EAttribute(container, E2EAttribute.button, i);

      // action
      fireEvent.click(button);

      // result
      await waitFor(
        () => {
          expect(button.lastChild).toHaveClass(
            `${classNamesButton[classNameButton].name}--${RIPPLE_EFFECT_MODIFICATOR}`,
          );
        },
        { timeout: 100 },
      );
    }
  });
});

describe('ButtonGroup props', () => {
  it('should pass children', () => {
    // before
    const { container } = render(<ButtonGroup>{children}</ButtonGroup>);

    // result
    for (let i = 0; i < contents.length; i++) {
      expect(
        getByE2EAttribute(container, E2EAttribute.button, i),
      ).toHaveTextContent(contents[i]);
    }
  });

  it('should pass className', () => {
    // before
    const { container } = render(
      <ButtonGroup className={className}>{children}</ButtonGroup>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.buttonGroup)).toHaveClass(
      className,
    );
  });

  it('should pass color', () => {
    // mock
    const color = ButtonColor.secondary;

    // before
    const { container } = render(
      <ButtonGroup color={color}>{children}</ButtonGroup>,
    );

    // result
    for (let i = 0; i < contents.length; i++) {
      expect(getByE2EAttribute(container, E2EAttribute.button, i)).toHaveClass(
        classNamesButton[classNameButton].modificators[color],
      );
    }
  });

  it('should pass disabled', () => {
    // before
    const { container } = render(
      <ButtonGroup disabled>{children}</ButtonGroup>,
    );

    // result
    for (let i = 0; i < contents.length; i++) {
      expect(
        getByE2EAttribute(container, E2EAttribute.button, i),
      ).toBeDisabled();
    }
  });

  it('should pass disableRippleEffect', async () => {
    // before
    const { container } = render(
      <ButtonGroup disabledRippleEffect>{children}</ButtonGroup>,
    );

    for (let i = 0; i < contents.length; i++) {
      // find
      const button = getByE2EAttribute(container, E2EAttribute.button, i);

      // action
      fireEvent.click(button);

      // result
      await waitFor(
        () => {
          expect(button.lastChild).toHaveTextContent(contents[i]);
        },
        { timeout: 100 },
      );
    }
  });

  it('should pass e2eAttribute', () => {
    // before
    const { container } = render(
      <ButtonGroup e2eAttribute={E2EAttribute.buttonGroup}>
        {children}
      </ButtonGroup>,
    );

    // result
    expect(
      getByE2EAttribute(container, E2EAttribute.buttonGroup),
    ).toHaveAttribute(getDataTestAttribute(E2EAttribute.buttonGroup));
  });

  it('should pass e2eValue', () => {
    // before
    const { container } = render(
      <ButtonGroup e2eValue="e2eValue">{children}</ButtonGroup>,
    );

    // result
    expect(
      getByE2EAttribute(container, E2EAttribute.buttonGroup),
    ).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.buttonGroup),
      'e2eValue',
    );
  });

  it('should pass forcedHover', () => {
    // before
    const { container } = render(
      <ButtonGroup forcedHover>{children}</ButtonGroup>,
    );

    // result
    for (let i = 0; i < contents.length; i++) {
      expect(getByE2EAttribute(container, E2EAttribute.button, i)).toHaveClass(
        classNamesButton[classNameButton].modificators.forcedHover,
      );
    }
  });

  it('should pass orientation', () => {
    // mock
    const orientations = enumToArray(
      ButtonGroupOrientation,
    ) as Array<ButtonGroupOrientation>;

    // before
    const { container } = render(
      <>
        {orientations.map((orientation) => (
          <ButtonGroup
            e2eValue={orientation}
            key={orientation}
            orientation={orientation}
          >
            {children}
          </ButtonGroup>
        ))}
      </>,
    );

    orientations.forEach((orientation) => {
      // find
      const buttonGroup = getByE2EAttribute(
        container,
        E2EAttribute.buttonGroup,
        orientation,
      );

      for (let i = 0; i < contents.length; i++) {
        // find
        const button = getByE2EAttribute(buttonGroup, E2EAttribute.button, i);

        // result
        expect(button).toHaveClass(
          classNamesButtonGroup.button.modificators[orientation],
        );
      }
    });
  });

  it('should pass size', () => {
    // mock
    const size = ButtonSize.large;

    // before
    const { container } = render(
      <ButtonGroup forcedHover size={size}>
        {children}
      </ButtonGroup>,
    );

    // result
    for (let i = 0; i < contents.length; i++) {
      expect(getByE2EAttribute(container, E2EAttribute.button, i)).toHaveClass(
        classNamesButton[classNameButton].modificators[size],
      );
    }
  });

  it('should pass style', () => {
    // before
    const { container } = render(
      <ButtonGroup style={{ width: '100%' }}>{children}</ButtonGroup>,
    );

    // result
    expect(
      getByE2EAttribute(container, E2EAttribute.buttonGroup),
    ).toHaveAttribute('style', 'width: 100%;');
  });

  it('should pass type', () => {
    // before
    const { container } = render(
      <ButtonGroup type="submit">{children}</ButtonGroup>,
    );

    // result
    for (let i = 0; i < contents.length; i++) {
      expect(
        getByE2EAttribute(container, E2EAttribute.button, i),
      ).toHaveAttribute('type', 'submit');
    }
  });

  it('should pass variant', () => {
    // mock
    const variant = ButtonVariant.outlined;

    // before
    const { container } = render(
      <ButtonGroup variant={variant}>{children}</ButtonGroup>,
    );

    // result
    for (let i = 0; i < contents.length; i++) {
      expect(getByE2EAttribute(container, E2EAttribute.button, i)).toHaveClass(
        classNamesButton[classNameButton].modificators[variant],
      );
    }
  });
});
