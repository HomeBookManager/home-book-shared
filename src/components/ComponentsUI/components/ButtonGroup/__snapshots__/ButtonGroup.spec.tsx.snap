// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup snapshots should render ButtonGroup 1`] = `
<DocumentFragment>
  <div
    class="ButtonGroup ButtonGroup--horizontal"
    data-test-buttongroup=""
  >
    <button
      class="ButtonGroup__button ButtonGroup__button--primary ButtonGroup__button--horizontal ButtonGroup__button--contained Button Button--primary Button--medium Button--contained"
      data-test-button="0"
      type="button"
    >
      Button 1
    </button>
    <button
      class="ButtonGroup__button ButtonGroup__button--primary ButtonGroup__button--horizontal ButtonGroup__button--contained Button Button--primary Button--medium Button--contained"
      data-test-button="1"
      type="button"
    >
      Button 2
    </button>
    <button
      class="ButtonGroup__button ButtonGroup__button--primary ButtonGroup__button--horizontal ButtonGroup__button--contained Button Button--primary Button--medium Button--contained"
      data-test-button="2"
      type="button"
    >
      Button 3
    </button>
  </div>
</DocumentFragment>
`;
